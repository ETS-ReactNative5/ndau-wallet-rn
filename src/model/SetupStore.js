import { Alert } from 'react-native';
import AppConstants from '../AppConstants';

class SetupStore {
  constructor() {
    if (!SetupStore.instance) {
      SetupStore.instance = this;
    }

    this._userId = '';
    this._numberOfAccounts = 0;
    this._qrCode = '';
    this._encryptionPassword = '';
    this._entropy = '';
    this._recoveryPhrase = '';
    this._shuffledMap = [];
    this._shuffledWord = [];
    this._addressType = AppConstants.MAINNET_ADDRESS;
    this._walletName = '';
  }

  set userId(userId) {
    this._userId = userId;
    this._printData();
  }

  get userId() {
    console.debug(`SetupStore.userId ${this._userId}`);
    return this._userId;
  }

  set numberOfAccounts(numberOfAccounts) {
    this._numberOfAccounts = numberOfAccounts;
    this._printData();
  }

  get numberOfAccounts() {
    console.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`);
    return this._numberOfAccounts;
  }

  set qrCode(qrCode) {
    this._qrCode = qrCode;
    this._printData();
  }

  get qrCode() {
    console.debug(`SetupStore.qrCode ${this._qrCode}`);
    return this._qrCode;
  }

  set encryptionPassword(encryptionPassword) {
    this._encryptionPassword = encryptionPassword;
    this._printData();
  }

  get encryptionPassword() {
    console.debug(`SetupStore.encryptionPassword ${this._encryptionPassword}`);
    return this._encryptionPassword;
  }

  set entropy(entropy) {
    this._entropy = entropy;
    this._printData();
  }

  get entropy() {
    console.debug(`SetupStore.entropy ${this._entropy}`);
    return this._entropy;
  }

  set recoveryPhrase(recoveryPhrase) {
    this._recoveryPhrase = recoveryPhrase.slice();
    this._printData();
  }

  get recoveryPhrase() {
    console.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`);
    return this._recoveryPhrase;
  }

  set shuffledWords(shuffledWords) {
    this._shuffledWords = shuffledWords.slice();
    this._printData();
  }

  get shuffledWords() {
    console.debug(`SetupStore.shuffledWords ${this._shuffledWords}`);
    return this._shuffledWords;
  }

  set shuffledMap(shuffledMap) {
    this._shuffledMap = shuffledMap.slice();
    this._printData();
  }

  get shuffledMap() {
    console.debug(`SetupStore.shuffledMap ${this._shuffledMap}`);
    return this._shuffledMap;
  }

  toggleAddressType() {
    const oldAddressType = this._addressType;
    const newAddressType =
      this._addressType === AppConstants.MAINNET_ADDRESS
        ? AppConstants.TESTNET_ADDRESS
        : AppConstants.MAINNET_ADDRESS;

    Alert.alert(
      'Information',
      `Old address type was ${oldAddressType} which has been moved to ${newAddressType}`,
      [{ text: 'OK', onPress: () => { } }],
      { cancelable: false }
    );

    this._addressType = newAddressType;
    this._printData();
  }

  get addressType() {
    console.debug(`SetupStore.addressType ${this._addressType}`);
    return this._addressType;
  }

  set walletName(walletName) {
    this._walletName = walletName;
    this._printData();
  }

  get walletName() {
    console.debug(`SetupStore.walletName ${this._walletName}`);
    return this._walletName;
  }

  _printData = () => {
    // console.debug(`SetupStore.userId ${this._userId}`);
    // console.debug(`SetupStore.numberOfAccounts ${this._numberOfAccounts}`);
    // console.debug(`SetupStore.qrCode ${this._qrCode}`);
    // console.debug(`SetupStore.entropy ${this._entropy}`);
    // console.debug(`SetupStore.recoveryPhrase ${this._recoveryPhrase}`);
    // console.debug(`SetupStore.shuffledWords ${this._shuffledWords}`);
    // console.debug(`SetupStore.shuffledMap ${this._shuffledMap}`);
    // console.debug(`SetupStore.walletName ${this._walletName}`);
    // console.debug(`SetupStore.addressType ${this._addressType}`);
  };
}

const instance = new SetupStore();

export default instance;
