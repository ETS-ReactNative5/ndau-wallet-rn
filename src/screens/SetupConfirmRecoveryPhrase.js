import React, { Component } from 'react'
import {
  PixelRatio,
  StyleSheet,
  View,
  Text,
  TouchableHighlight
} from 'react-native'
import groupIntoRows from '../helpers/groupIntoRows'
import SetupStore from '../stores/SetupStore'
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp
} from 'react-native-responsive-screen'
import EntropyHelper from '../helpers/EntropyHelper'
import FlashNotification from '../components/common/FlashNotification'
import MultiSafeHelper from '../helpers/MultiSafeHelper'
import DataFormatHelper from '../helpers/DataFormatHelper'
import UserStore from '../stores/UserStore'
import AppConstants from '../AppConstants'
import LoggingService from '../services/LoggingService'
import {
  SetupContainer,
  RecoveryPhraseConfirmation,
  RecoveryPhraseConfirmationButtons
} from '../components/setup'
import { LargeButtons, ParagraphText } from '../components/common'

const _ = require('lodash')

const MAX_ERRORS = 4 // 4 strikes and you're out
const DEFAULT_ROW_LENGTH = 4 // 3 items per row
const TOTAL_RECOVERY_WORDS = 12

let boxWidth = '25%'
let boxHeight = '10%'

class SetupConfirmRecoveryPhrase extends Component {
  constructor (props) {
    super(props)
    this.state = {
      textColor: '#ffffff',
      inError: false,
      mustRetry: false,
      errorCount: 0,
      match: false,
      selected: []
    }

    // chop the words into DEFAULT_ROW_LENGTH-tuples
    this.rowLength = DEFAULT_ROW_LENGTH
    if (PixelRatio.getFontScale() > 2) {
      this.rowLength = 1
      boxWidth = '100%'
      boxHeight = '15%'
    }
  }

  showNextSetup = async () => {
    let user = UserStore.getUser()
    if (user) {
      // if a user is present then we have wallets and can assume
      // they are logged in, so we get the password setup
      const password = await UserStore.getPassword()
      user = await MultiSafeHelper.addNewWallet(
        user,
        DataFormatHelper.convertRecoveryArrayToString(
          SetupStore.recoveryPhrase
        ),
        AppConstants.TEMP_ID,
        user.userId,
        SetupStore.numberOfAccounts,
        password
      )
    }

    LoggingService.debug(
      `user going into SetupWalletName: ${JSON.stringify(user, null, 2)}`
    )

    this.props.navigation.navigate('SetupWalletName')
  }

  pushBack = async () => {
    const user = UserStore.getUser()

    await EntropyHelper.generateEntropy()
    this.props.navigation.navigate('SetupYourWallet')
    FlashNotification.hideMessage()
  }

  render () {
    const shuffledWords = SetupStore.shuffledWords
    const words = groupIntoRows(shuffledWords, this.rowLength)

    const selectedWords = this.state.selected.map(selectedIndex => {
      return shuffledWords[selectedIndex]
    })
    const formattedSelectedWords = groupIntoRows(selectedWords, this.rowLength)

    // lookup table for word highlights
    const selected = this.state.selected.reduce((arr, cur) => {
      arr[cur] = true
      return arr
    }, {})

    if (this.state.selected.length === TOTAL_RECOVERY_WORDS) {
      return (
        <SetupContainer {...this.props} pageNumber={15}>
          <ParagraphText>Please verify your recovery phrase.</ParagraphText>
          <RecoveryPhraseConfirmation
            words={formattedSelectedWords || []}
            rowTextView={styles.rowTextView}
          />
          <LargeButtons
            text='Is this correct?'
            onPress={() => this.showNextSetup()}
          >
            Confirm
          </LargeButtons>
          <LargeButtons bottom secondary onPress={() => this.pushBack()}>
            Back
          </LargeButtons>
        </SetupContainer>
      )
    } else {
      return (
        <SetupContainer {...this.props} pageNumber={15}>
          <ParagraphText>Please verify your recovery phrase.</ParagraphText>
          <RecoveryPhraseConfirmation
            words={formattedSelectedWords || []}
            rowTextView={styles.rowTextView}
          />
          <RecoveryPhraseConfirmationButtons
            words={words}
            errorWord={this.state.errorWord}
            selected={selected}
            rowTextView={styles.rowTextView}
            handleClick={this.handleClick}
          />
        </SetupContainer>
      )
    }
  }

  compare (correctSoFar, selected) {
    if (_.isEqual(correctSoFar, selected)) {
      return true
    } else {
      // compare strings
      const recoveryPhrase = SetupStore.shuffledWords

      return correctSoFar.every(index => {
        const correctIndex = correctSoFar[index]
        const selectedIndex = selected[index]
        const correctWord = recoveryPhrase[correctIndex]
        const selectedWord = recoveryPhrase[selectedIndex]
        const wordsMatch =
          correctWord && correctWord.length > 0 && correctWord === selectedWord

        return wordsMatch
      })
    }
  }

  checkMistakes () {
    const { selected } = this.state
    const correctSoFar = SetupStore.shuffledMap.slice(
      0,
      this.state.selected.length
    )

    if (!this.compare(correctSoFar, selected)) {
      const errorText = this.state.mustRetry
        ? 'Please click the Back button to generate a new recovery phrase. Write down your phrase instead of memorizing it, or you may lose access to your ndau.'
        : 'Please enter the words in the correct order. De-select the last word to continue.'
      FlashNotification.showError(errorText)

      let errorCount = this.state.errorCount + 1
      this.setState({
        inError: true,
        mustRetry: errorCount >= MAX_ERRORS,
        errorCount: errorCount,
        errorWord: this.state.selected[this.state.selected.length - 1]
      })
    } else {
      this.setState({
        inError: false,
        errorWord: null
      })
      FlashNotification.hideMessage()
    }
  }

  checkDone () {
    if (_(this.state.selected).isEqual(SetupStore.shuffledMap)) {
      this.setState({ match: true })
    }
  }

  handleClick = index => {
    const selected = this.state.selected.slice()
    const foundIndex = selected.indexOf(index)
    if (foundIndex !== -1) {
      // already selected item was clicked
      selected.splice(foundIndex, 1)
      this.setState({ selected }, this.afterClick)
    } else if (!this.state.inError) {
      selected.push(index)
      this.setState({ selected }, this.afterClick)
    }
  }
  afterClick () {
    this.checkMistakes()
    this.checkDone()
  }
}

function Word (props) {
  let bgColor = 'transparent'
  if (props.error) {
    bgColor = AppConstants.WARNING_ICON_COLOR
  } else if (props.selected) {
    bgColor = AppConstants.SQUARE_BUTTON_COLOR
  }

  return (
    <TouchableHighlight onPress={props.onPress}>
      <View
        style={{
          height: hp(boxHeight),
          width: wp(boxWidth),
          backgroundColor: bgColor,
          alignItems: 'center',
          justifyContent: 'center',
          flex: 1,
          borderRadius: 3,
          marginTop: hp('1%'),
          marginBottom: hp('1%'),
          marginLeft: wp('1%'),
          marginRight: wp('1%')
        }}
      >
        <Text
          style={{
            color: '#ffffff',
            fontSize: 20,
            fontFamily: 'TitilliumWeb-Regular',
            textAlign: 'center'
          }}
        >
          {props.children}
        </Text>
      </View>
    </TouchableHighlight>
  )
}

const styles = StyleSheet.create({
  navButtons: {
    width: wp('40%')
  },
  rowView: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-evenly'
  }
})

export default SetupConfirmRecoveryPhrase
