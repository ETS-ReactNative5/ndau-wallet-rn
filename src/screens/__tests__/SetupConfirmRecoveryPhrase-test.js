import React from 'react';
import { StyleSheet } from 'react-native';
import SetupConfirmRecoveryPhrase from '../SetupConfirmRecoveryPhrase';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import SetupStore from '../../model/SetupStore';

function makeStyles() {
  return StyleSheet.create({
    wizardText: {
      color: '#ffffff',
      fontSize: 20
    }
  });
}

function makeNavigator() {
  return {
    navigate: () => {}
  };
}

function makeWords() {
  return 'zero one two three four four six seven eight nine ten eleven'.split(' ');
}

describe('SetupConfirmRecoveryPhrase presentation', () => {
  beforeEach(() => {
    SetupStore.recoveryPhrase = makeWords();
    SetupStore.entropy = 'dGVzdGluZ3dlc3Rpbmdh';
    SetupStore.shuffledMap = [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 ];
    SetupStore.shuffledWords = makeWords();
    this.tree = renderer
      .create(
        <SetupConfirmRecoveryPhrase navigation={makeNavigator()} parentStyles={makeStyles()} />
      )
      .toJSON();
  });

  test('renders correctly', () => {
    expect(this.tree).toMatchSnapshot();
  });
});

describe('SetupConfirmRecoveryPhrase behavior', () => {
  beforeEach(() => {
    // set up default wrapper for every test.
    SetupStore.recoveryPhrase = makeWords();
    SetupStore.entropy = 'dGVzdGluZ3dlc3Rpbmdh';
    SetupStore.shuffledMap = [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 ];
    SetupStore.shuffledWords = makeWords();
    this.wrapper = mount(<SetupConfirmRecoveryPhrase navigation={makeNavigator()} />);
    this.press = (index) => {
      this.wrapper.find('Word').at(index).props().onPress();
    };
  });

  test('done button starts disabled', () => {
    const doneButton = this.wrapper.find('Button').at(1);
    expect(doneButton.prop('disabled')).toBeTruthy();
  });

  test('trigger error state with incorrect sequence', () => {
    this.press(0);
    this.press(1);
    this.press(2);
    this.press(4); // wrong
    this.wrapper.update();
    expect(this.wrapper.state().inError).toBeTruthy();
  });

  test('error count goes up with incorrect input', () => {
    this.press(4); // wrong
    this.wrapper.update();
    expect(this.wrapper.state().errorWord).toBe(4);
    expect(this.wrapper.state().errorCount).toBe(1);
  });

  test('done button is enabled after correct sequence', () => {
    makeWords().forEach((e, i) => {
      this.wrapper.find('Word').at(i).props().onPress();
    });
    this.wrapper.update();
    const doneButton = this.wrapper.find('Button').at(1);
    expect(doneButton.prop('disabled')).toBeFalsy();
  });

  test('trigger error state with incorrect word', () => {
    this.press(6); // wrong
    this.wrapper.update();
    expect(this.wrapper.state().inError).toBeTruthy();
    expect(this.wrapper.state().errorWord).toBe(6);
    expect(this.wrapper.state().errorCount).toBe(1);
  });

  test('correct error state by deselecting word', () => {
    this.press(6); // wrong
    this.press(6); // correction
    this.wrapper.update();
    expect(this.wrapper.state().inError).toBeFalsy();
    expect(this.wrapper.state().errorCount).toBe(1);
  });

  it('should not be able to click any buttons after a mistake', () => {
    this.press(6); // wrong
    this.press(1); // another wrong
    this.wrapper.update();
    expect(this.wrapper.state().selected).toEqual([ 6 ]);
    expect(this.wrapper.state().errorCount).toBe(1);
  });

  test('trigger error state with incorrect sequence', () => {
    this.press(0);
    this.press(1);
    this.press(2);
    this.press(4); // wrong
    this.wrapper.update();
    expect(this.wrapper.state().inError).toBeTruthy();
    expect(this.wrapper.state().errorWord).toBe(4);
    expect(this.wrapper.state().errorCount).toBe(1);
  });

  test('lockem out after too many mistakes', () => {
    this.press(1); // wrong
    expect(this.wrapper.state().errorCount).toBe(1);
    this.press(1); // correction

    this.press(2); // wrong
    expect(this.wrapper.state().errorCount).toBe(2);
    this.press(2); // correction

    this.press(0); // correct

    this.press(4); // wrong
    expect(this.wrapper.state().errorCount).toBe(3);
    this.press(4); // correction

    this.press(4); // wrong again

    expect(this.wrapper.state().errorCount).toBe(4);
    this.wrapper.update();
    expect(this.wrapper.state().mustRetry).toBeTruthy();
  });

  test('allow incorrect sequence of duplicate words', () => {
    this.press(0);
    this.press(1);
    this.press(2);
    this.press(3);
    this.press(5); // "four"
    this.press(4); // "four"
    this.wrapper.update();
    expect(this.wrapper.state().inError).toBeFalsy();
  });
});
