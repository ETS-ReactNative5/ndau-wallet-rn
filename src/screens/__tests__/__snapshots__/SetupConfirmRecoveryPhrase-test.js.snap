// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetupConfirmRecoveryPhrase presentation renders correctly 1`] = `
<View
  onLayout={[Function]}
  pointerEvents="box-none"
  style={
    Object {
      "backgroundColor": "#1c2227",
      "flex": 1,
      "paddingBottom": 0,
      "paddingLeft": 0,
      "paddingRight": 0,
      "paddingTop": 20,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1c2227",
        "flex": 1,
        "paddingBottom": 26.5,
        "paddingLeft": 15,
        "paddingRight": 15,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "marginBottom": 93.5,
            }
          }
        />
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 20,
              }
            }
          >
            To confirm that you recorded the phrase, tap the words below in order.
             
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                zero
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                one
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                two
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                three
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                four
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                four
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                six
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                seven
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                eight
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                nine
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                ten
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderRadius": 3,
                  "flex": 1,
                  "height": 133.5,
                  "justifyContent": "center",
                  "marginBottom": 7.5,
                  "marginTop": 7.5,
                  "width": 225,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "TitilliumWeb-Regular",
                    "fontSize": 20,
                    "textAlign": "center",
                  }
                }
              >
                eleven
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "display": "flex",
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        style={
          Object {
            "justifyContent": "space-between",
            "marginTop": 26.5,
          }
        }
      >
        <View
          accessibilityComponentType="button"
          accessibilityTraits="button"
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignContent": "center",
              "backgroundColor": "#4e957a",
              "borderRadius": 3,
              "height": 93.5,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "alignSelf": "center",
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Back (resets phrase)
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "justifyContent": "space-between",
            "marginTop": 26.5,
          }
        }
      >
        <View
          accessibilityComponentType="button"
          accessibilityTraits="button"
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignContent": "center",
              "backgroundColor": "#696969",
              "borderRadius": 3,
              "height": 93.5,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "alignSelf": "center",
                "color": "#ffffff",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Next
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
